pypulseq/Sequence/sequence.py

```diff
@@ -1510,11 +1511,15 @@ def waveforms_and_times(
                             :, wave_cnt[j] + np.arange(length - 1)
                         ] = wave_data_local[:, 1:]
                         wave_cnt[j] += length - 1
-                    if wave_cnt[j] != len(np.unique(wave_data[j][0, : wave_cnt[j]])):
+
+                    rftdiff = np.diff(wave_data[j][0, : wave_cnt[j]])

```

Comment:
Request you to try `list(set(...))`, is that faster?

Reply:
I tried that, it is extremely slow.

As an aside, I think we should not do this check here, but rather during RF creation, for the following reasons:

1. It does not make sense to for the sequence developer to create an RF object with non-monotonically increasing time points in the first place. `waveforms_and_times()` function may not be called typically, so this issue may not be caught in this case.

2. We create RF object once, and add it many times (i.e. every TR). This way, we are checking the same RF pulse many times unnecessarily, hence the slow-down.

I can also ask Maxim about this design choice. But if you are convinced with my argument, I can remove this check and put it inside make RF functions.

Reply:
Discussion with Maxim: [https://github.com/pulseq/pulseq/issues/25](https://github.com/pulseq/pulseq/issues/25)

TL;DR: He is planning to make some changes to reduce the calls to unique(). Till then, I am leaving it at diff.